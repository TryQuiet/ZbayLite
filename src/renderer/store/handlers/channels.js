import Immutable from 'immutable'
import { DateTime } from 'luxon'
import { createAction, handleActions } from 'redux-actions'
import BigNumber from 'bignumber.js'
import { remote } from 'electron'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification,
  LoaderState
} from './utils'
import notificationsHandlers from './notifications'
import channelHandlers from './channel'
import identitySelectors from '../selectors/identity'
import channelsSelectors from '../selectors/channels'
import channelSelectors from '../selectors/channel'
import modalsHandlers from './modals'
import { messages } from '../../zbay'
import client from '../../zcash'
import logsHandlers from '../../store/handlers/logs'
import { networkFee, actionTypes } from '../../../shared/static'
import history from '../../../shared/history'
import electronStore from '../../../shared/electronStore'
import contactsHandlers from './contacts'
import ownedChannelsHandlers from './ownedChannels'

const toBigNumber = x => new BigNumber(x)

export const ChannelsState = Immutable.Record(
  {
    data: Immutable.List(),
    loader: LoaderState({ loading: true })
  },
  'ChannelsState'
)

export const initialState = ChannelsState()

const loadChannels = createAction(
  actionTypes.LOAD_IDENTITY_CHANNELS,
  async id => {
    return ''
  }
)
const loadChannelsToNode = createAction(
  actionTypes.LOAD_IDENTITY_CHANNELS,
  async (id, isNewUser) => {
    return ''
  }
)

const setLastSeen = createAction(actionTypes.SET_CHANNELS_LAST_SEEN)
const setDescription = createAction(actionTypes.SET_CHANNEL_DESCRIPTION)
const setUnread = createAction(actionTypes.SET_CHANNEL_UNREAD)
const setShowInfoMsg = createAction(actionTypes.SET_SHOW_INFO_MSG)
const setAdvertFee = createAction(actionTypes.SET_ADVERT_FEE)
const setOnlyRegistered = createAction(actionTypes.SET_ONLY_REGISTERED)

export const actions = {
  loadChannels,
  setLastSeen,
  setUnread,
  loadChannelsToNode,
  setDescription,
  setAdvertFee,
  setOnlyRegistered
}

const _createChannel = async values => {
  const address = await client.getNewShieldedAdress()
  return address
}

const createChannel = (values, formActions, setStep) => async (
  dispatch,
  getState
) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  const balance = identitySelectors.balance('zec')(getState())
  const signerPubKey = identitySelectors.signerPubKey(getState())
  try {
    if (balance.lt(networkFee)) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({
            message: `You need minimum ${networkFee} ZEC to create a channel. `
          })
        )
      )
      formActions.setSubmitting(false)
      return
    }
    setStep(1)
    const address = await _createChannel(values)
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Creating channel ${address}`
      })
    )
    const result = await dispatch(
      channelHandlers.epics.sendChannelSettingsMessage({ address: address })
    )
    if (result === -1) {
      setStep(0)
      dispatch(closeModal())
      return
    }
    const importedChannels = electronStore.get(`importedChannels`) || {}
    const viewingKey = await client.getViewingKey(address)
    electronStore.set('importedChannels', {
      ...importedChannels,
      [address]: {
        address: address,
        name: values.name,
        description: '',
        owner: signerPubKey,
        keys: { ivk: viewingKey }
      }
    })
    await dispatch(
      contactsHandlers.actions.addContact({
        key: address,
        contactAddress: address,
        username: values.name
      })
    )
    dispatch(ownedChannelsHandlers.epics.getOwnedChannels())
    history.push(`/main/channel/${address}`)
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `Successfully created ${values.name} channel.`
        })
      )
    )
    setStep(0)
    dispatch(closeModal())
  } catch (error) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: `Failed to create channel: ${error.message}`
        })
      )
    )
    setStep(0)
    formActions.setSubmitting(false)
  }
}

const withdrawMoneyFromChannels = () => async (dispatch, getState) => {
  const ownedChannels = channelsSelectors.ownedChannels(getState())
  let earnedAmount = toBigNumber(0)
  for (const channel of ownedChannels) {
    const channelAddress = channel.get('address')
    const amount = await getMoneyFromChannel(channelAddress)(dispatch, getState)
    earnedAmount = earnedAmount.plus(amount)
  }
  if (earnedAmount.gt(toBigNumber(0))) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `You will shortly receive ${earnedAmount.toFixed(
            4
          )} Zcash for commission from your created channels.`
        })
      )
    )
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Creating new transfer with received money from channels`
      })
    )
  }
}

const getMoneyFromChannel = address => async (dispatch, getState) => {
  const amount = await client.accounting.balance(address)
  const identityAddress = identitySelectors.address(getState())
  if (amount.gt(0.0005)) {
    const transfer = messages.createEmptyTransfer({
      address: identityAddress,
      amount: amount.minus(networkFee).toString(),
      identityAddress: address
    })
    await client.payment.send(transfer)
    return amount.minus(networkFee)
  }
  return toBigNumber(0)
}

const updateLastSeen = ({ channelId }) => async (dispatch, getState) => {
  const lastSeen = DateTime.utc()
  const unread = channelSelectors.unread(getState())
  remote.app.badgeCount = remote.app.badgeCount - unread
  dispatch(setLastSeen({ channelId, lastSeen }))
  dispatch(setUnread({ channelId, unread: 0 }))
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: `Updating last seen ${channelId}`
    })
  )
}
const updateSettings = ({ channelId, time, data }) => async (
  dispatch,
  getState
) => {
  const lastSeen = channelsSelectors.lastSeen(channelId)(getState())
  const timeOfTransaction = DateTime.fromSeconds(time)
  if (timeOfTransaction.plus({ minutes: 2 }) > lastSeen) {
    dispatch(updateShowInfoMsg(true))
  }
  dispatch(
    setDescription({ channelId, description: data.updateChannelDescription })
  )
  dispatch(
    setOnlyRegistered({
      channelId,
      onlyRegistered: !!parseInt(data.updateOnlyRegistered)
    })
  )
  dispatch(setAdvertFee({ channelId, advertFee: data.updateMinFee }))
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: `Updating channel settings`
    })
  )
}

const updateShowInfoMsg = showInfoMsg => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  dispatch(setShowInfoMsg({ channelId, showInfoMsg }))
}

export const epics = {
  createChannel,
  updateLastSeen,
  getMoneyFromChannel,
  withdrawMoneyFromChannels,
  updateShowInfoMsg,
  updateSettings
}

export const reducer = handleActions(
  {
    [typePending(actionTypes.LOAD_IDENTITY_CHANNELS)]: state =>
      state
        .setIn(['loader', 'loading'], true)
        .setIn(['loader', 'message'], 'Loading channels'),
    [typeFulfilled(actionTypes.LOAD_IDENTITY_CHANNELS)]: (
      state,
      { payload: data }
    ) =>
      state
        .set('data', Immutable.fromJS(data))
        .setIn(['loader', 'loading'], false),
    [typeRejected(actionTypes.LOAD_IDENTITY_CHANNELS)]: (
      state,
      { payload: error }
    ) => state.setIn(['loader', 'loading'], false),
    [setDescription]: (state, { payload: { channelId, description } }) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch =>
        ch.set('description', description)
      )
    },
    [setLastSeen]: (state, { payload: { channelId, lastSeen } }) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch => ch.set('lastSeen', lastSeen))
    },
    [setOnlyRegistered]: (
      state,
      { payload: { channelId, onlyRegistered } }
    ) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch =>
        ch.set('onlyRegistered', onlyRegistered)
      )
    },
    [setAdvertFee]: (state, { payload: { channelId, advertFee } }) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch =>
        ch.set('advertFee', parseFloat(advertFee))
      )
    },
    [setShowInfoMsg]: (state, { payload: { channelId, showInfoMsg } }) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch =>
        ch.set('showInfoMsg', showInfoMsg)
      )
    },
    [setUnread]: (state, { payload: { channelId, unread } }) => {
      const index = state.data.findIndex(
        channel => channel.get('id') === channelId
      )
      return state.updateIn(['data', index], ch => ch.set('unread', unread))
    }
  },
  initialState
)

export default {
  actions,
  epics,
  reducer
}
