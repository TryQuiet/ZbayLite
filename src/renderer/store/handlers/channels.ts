import { produce, immerable } from 'immer'
import { DateTime } from 'luxon'
import { createAction, handleActions, Action } from 'redux-actions'
import BigNumber from 'bignumber.js'
import { remote } from 'electron'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification
} from './utils'
import notificationsHandlers from './notifications'
import channelHandlers from './channel'
import identitySelectors from '../selectors/identity'
import channelsSelectors from '../selectors/channels'
import publicChannelsSelectors from '../selectors/publicChannels'
import channelSelectors from '../selectors/channel'
import modalsHandlers from './modals'
import { messages } from '../../zbay'
import client from '../../zcash'
import { networkFee, actionTypes } from '../../../shared/static'
import history from '../../../shared/history'
import electronStore from '../../../shared/electronStore'
import contactsHandlers from './contacts'
import ownedChannelsHandlers from './ownedChannels'
import contactsSelectors from '../selectors/contacts'
import { publicChannelsActions } from '../../sagas/publicChannels/publicChannels.reducer'

import { ActionsType, PayloadType } from './types'

const toBigNumber = x => new BigNumber(x)

// TODO: remove after changing in tests
export const ChannelsState = {
  data: [],
  loader: {
    loading: false,
    message: ''
  }
}
interface Channel {
  id: string
  description: string
  lastSeen: DateTime
  onlyRegistered: boolean
  advertFee: number
  showInfoMsg: boolean
  unread: number
}
class Channels {
  data: Channel[]
  loader: {
    loading: boolean
    message: string
  }

  constructor(values?: Partial<Channels>) {
    Object.assign(this, values)
    this[immerable] = true
  }
}

export const initialState: Channels = new Channels({
  data: [],
  loader: {
    loading: false,
    message: ''
  }
})

const loadChannels = createAction(actionTypes.LOAD_IDENTITY_CHANNELS, async () => {
  return ''
})

const loadChannelsToNode = createAction(actionTypes.LOAD_IDENTITY_CHANNELS, async () => {
  return ''
})

export type ChannelsStore = Channels

const setLastSeen = createAction<{ channelId: string; lastSeen: DateTime }>(
  actionTypes.SET_CHANNELS_LAST_SEEN
)
const setDescription = createAction<{ channelId: string; description: string }>(
  actionTypes.SET_CHANNEL_DESCRIPTION
)
const setUnread = createAction<{ unread: number; channelId: string }>(
  actionTypes.SET_CHANNEL_UNREAD
)
const setShowInfoMsg = createAction<{ channelId: string; showInfoMsg: boolean }>(
  actionTypes.SET_SHOW_INFO_MSG
)
const setAdvertFee = createAction<{ channelId: string; advertFee: string }>(
  actionTypes.SET_ADVERT_FEE
)
const setOnlyRegistered = createAction<{ channelId: string; onlyRegistered: boolean }>(
  actionTypes.SET_ONLY_REGISTERED
)

export const actions = {
  loadChannels,
  setLastSeen,
  setUnread,
  loadChannelsToNode,
  setDescription,
  setAdvertFee,
  setOnlyRegistered,
  setShowInfoMsg
}

export type ChannelsActions = ActionsType<typeof actions>

const _createChannel = async () => {
  const address = await client.getNewShieldedAdress()
  return address
}

const createChannel = (values, formActions, setStep) => async (dispatch, getState) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  const balance = identitySelectors.balance('zec')(getState())
  const signerPubKey = identitySelectors.signerPubKey(getState())
  try {
    if (balance.lt(networkFee)) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({
            message: `You need minimum ${networkFee} ZEC to create a channel. `
          })
        )
      )
      formActions.setSubmitting(false)
      return
    }
    setStep(1)
    const address = await _createChannel()
    const result = await dispatch(
      channelHandlers.epics.sendChannelSettingsMessage({ address: address })
    )
    if (result === -1) {
      setStep(0)
      dispatch(closeModal())
      return
    }
    const importedChannels = electronStore.get('importedChannels') || {}
    const viewingKey = await client.getViewingKey(address)
    electronStore.set('importedChannels', {
      ...importedChannels,
      [address]: {
        address: address,
        name: values.name,
        description: '',
        owner: signerPubKey,
        keys: { ivk: viewingKey }
      }
    })
    await dispatch(
      contactsHandlers.actions.addContact({
        key: address,
        contactAddress: address,
        username: values.name
      })
    )
    dispatch(ownedChannelsHandlers.epics.getOwnedChannels())
    history.push(`/main/channel/${address}`)
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `Successfully created ${values.name} channel.`
        })
      )
    )
    setStep(0)
    dispatch(closeModal())
  } catch (error) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: `Failed to create channel: ${error.message}`
        })
      )
    )
    setStep(0)
    formActions.setSubmitting(false)
  }
}

const subscribeForPublicChannels = () => async (dispatch, getState) => {
  /** Subscribe for public channels from contacts (joined public channels) */
  const publicChannelsContacts = contactsSelectors.publicChannelsContacts(getState())
  for (const publicChannel of publicChannelsContacts) {
    const channel = publicChannelsSelectors.publicChannelsByName(publicChannel.username)(getState())
    console.log('subscribing for ', channel.name)
    if (channel) {
      dispatch(publicChannelsActions.subscribeForTopic(channel))
    }
  }
}

const withdrawMoneyFromChannels = () => async (dispatch, getState) => {
  const ownedChannels = channelsSelectors.ownedChannels(getState())
  let earnedAmount = toBigNumber(0)
  for (const channel of ownedChannels) {
    const channelAddress = channel.get('address')
    const amount = await getMoneyFromChannel(channelAddress)(dispatch, getState)
    earnedAmount = earnedAmount.plus(amount)
  }
  if (earnedAmount.gt(toBigNumber(0))) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: `You will shortly receive ${earnedAmount.toFixed(
            4
          )} Zcash for commission from your created channels.`
        })
      )
    )
  }
}

const getMoneyFromChannel = address => async (_dispatch, getState) => {
  const amount = await client.accounting.balance(address)
  const identityAddress = identitySelectors.address(getState())
  if (amount.gt(0.0005)) {
    const transfer = messages.createEmptyTransfer({
      address: identityAddress,
      amount: amount.minus(networkFee).toString(),
      identityAddress: address
    })
    await client.payment.send(transfer)
    return amount.minus(networkFee)
  }
  return toBigNumber(0)
}

const updateLastSeen = ({ channelId }) => async (dispatch, getState) => {
  const lastSeen = DateTime.utc()
  const unread = channelSelectors.unread(getState())
  remote.app.badgeCount = remote.app.badgeCount - unread
  dispatch(setLastSeen({ channelId, lastSeen }))
  dispatch(setUnread({ channelId, unread: 0 }))
}
const updateSettings = ({ channelId, time, data }) => async (dispatch, getState) => {
  const lastSeen = channelsSelectors.lastSeen(channelId)(getState())
  const timeOfTransaction = DateTime.fromSeconds(time)
  if (timeOfTransaction.plus({ minutes: 2 }) > lastSeen) {
    dispatch(updateShowInfoMsg(true))
  }
  dispatch(setDescription({ channelId, description: data.updateChannelDescription }))
  dispatch(
    setOnlyRegistered({
      channelId,
      onlyRegistered: !!parseInt(data.updateOnlyRegistered)
    })
  )
  dispatch(setAdvertFee({ channelId, advertFee: data.updateMinFee }))
}

const updateShowInfoMsg = showInfoMsg => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  dispatch(setShowInfoMsg({ channelId, showInfoMsg }))
}

export const epics = {
  createChannel,
  updateLastSeen,
  getMoneyFromChannel,
  withdrawMoneyFromChannels,
  updateShowInfoMsg,
  updateSettings,
  subscribeForPublicChannels
}

export const reducer = handleActions<ChannelsStore, PayloadType<ChannelsActions>>(
  {
    [typePending(actionTypes.LOAD_IDENTITY_CHANNELS)]: state =>
      produce(state, draft => {
        draft.loader.loading = true
        draft.loader.message = 'Loading channel'
      }),
    [typeFulfilled(actionTypes.LOAD_IDENTITY_CHANNELS)]: (state, { payload: data }: Action<any>) =>
      produce(state, draft => {
        draft.data = data
        draft.loader.loading = false
      }),
    [typeRejected(actionTypes.LOAD_IDENTITY_CHANNELS)]: state =>
      produce(state, draft => {
        draft.loader.loading = false
      }),
    [setDescription.toString()]: (
      state,
      { payload: { channelId, description } }: ChannelsActions['setDescription']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].description = description
      }),
    [setLastSeen.toString()]: (
      state,
      { payload: { channelId, lastSeen } }: ChannelsActions['setLastSeen']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].lastSeen = lastSeen
      }),
    [setOnlyRegistered.toString()]: (
      state,
      { payload: { channelId, onlyRegistered } }: ChannelsActions['setOnlyRegistered']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].onlyRegistered = onlyRegistered
      }),
    [setAdvertFee.toString()]: (
      state,
      { payload: { channelId, advertFee } }: ChannelsActions['setAdvertFee']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].advertFee = parseFloat(advertFee)
      }),
    [setShowInfoMsg.toString()]: (
      state,
      { payload: { channelId, showInfoMsg } }: ChannelsActions['setShowInfoMsg']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].showInfoMsg = showInfoMsg
      }),
    [setUnread.toString()]: (
      state,
      { payload: { channelId, unread } }: ChannelsActions['setUnread']
    ) =>
      produce(state, draft => {
        const index = state.data.findIndex(channel => channel.id === channelId)
        draft.data[index].unread = unread
      })
  },
  initialState
)

export default {
  actions,
  epics,
  reducer
}
