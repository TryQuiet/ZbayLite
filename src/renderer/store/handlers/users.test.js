/* eslint import/first: 0 */
jest.mock('../../zcash')

import Immutable from 'immutable'
import * as R from 'ramda'

import handlers, { initialState, ReceivedUser } from './users'
import { NodeState } from './node'
import selectors from '../selectors/users'
import { ChannelsState } from './channels'
import { initialState as IdentityState } from './identity'
import create from '../create'
import testUtils from '../../testUtils'

describe('users reducer', () => {
  let store = null

  const userChannel = [testUtils.channels.createUsersChannel(1)]

  beforeEach(() => {
    store = create({
      initialState: {
        node: { ...NodeState, isTestnet: true },
        identity: {
          ...IdentityState,
          data: {
            name: 'Saturn',
            id: 'test-id',
            address:
              'ztestsapling1mwmpvwy2aah7dlt0c9l47gde982xv6snxq2srddfzx8efn2qht6mxnz65rtst0426gtxje0eqlm',
            signerPrivKey: '4e577989a286937f565cea6426bd75fc0c3b166d7bd28d8357ba633b3736a41d',
            signerPubKey: '033ec8436690fc8313202e16a531b2ec4799dd91a33025357a5386979f3cf081af',
            transparentAddress: 'transparent-test-address'
          }
        },
        users: initialState,
        channels: {
          ...ChannelsState,
          data: userChannel
        }
      }
    })
    jest.clearAllMocks()
  })

  describe('handles actions -', () => {
    describe('setUsers', () => {
      it('sets users when empty', async () => {
        const users = await R.range(0, 3)
          .map(id => testUtils.messages.createReceivedUserMessage({ id: `test-user-${id}` }))
          .reduce(async (acc = Promise.resolve(Immutable.Map({})), message) => {
            const accumulator = await acc
            const user = ReceivedUser(message, accumulator)

            return Promise.resolve(accumulator.merge(user))
          }, Promise.resolve(Immutable.Map({})))
        const users2 = await R.range(3, 5)
          .map(id => testUtils.messages.createReceivedUserMessage({ id: `test-user-${id}` }))
          .reduce(async (acc = Promise.resolve(Immutable.Map({})), message) => {
            const accumulator = await acc
            const user = ReceivedUser(message, accumulator)

            return Promise.resolve(accumulator.merge(user))
          }, Promise.resolve(Immutable.Map({})))

        store.dispatch(handlers.actions.setUsers({ users }))
        store.dispatch(handlers.actions.setUsers({ messages: users2 }))

        expect(selectors.users(store.getState())).toMatchSnapshot()
      })
      it('sets users with same usernames', async () => {
        const users = await R.range(0, 3)
          .map(id => testUtils.messages.createReceivedUserMessage({ id: `test-user-${id}` }))
          .reduce(async (acc = Promise.resolve(Immutable.Map({})), message) => {
            const accumulator = await acc
            const user = ReceivedUser(message, accumulator)

            return Promise.resolve(accumulator.merge(user))
          }, Promise.resolve(Immutable.Map({})))
        store.dispatch(handlers.actions.setUsers({ users }))
        expect(selectors.users(store.getState())).toMatchSnapshot()
      })
    })
  })

  describe('handles epics -', () => {
    describe('isNicknameTaken', () => {
      beforeEach(() => {
        store = create({
          initialState: {
            node: { ...NodeState, isTestnet: true },
            users: {
              '02c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2392': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-1',
                address:
                  'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhyb4c'
              },
              '02c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c23vb': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-1',
                address:
                  'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhyb4c'
              },
              '04c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2355': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-2',
                address:
                  'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhylnn'
              },
              '04c9b30ea203dcd5776d014e0062a3232c00b74273094ebb1f119fb5cee88c2vbf': {
                firstName: 'testname',
                lastName: 'testlastname',
                nickname: 'test-user-2',
                address:
                  'ztestsapling1k059n2xjz5apmu49ud6xa0g4lywetd0zgpz2txe9xs5pu27fjjnp7c9yvtkcqlwz0n7qxrhylnn'
              }
            }
          }
        })
        jest.clearAllMocks()
      })

      it('should return true for taken username', async () => {
        const isNicknameTaken = await store.dispatch(handlers.epics.isNicknameTaken('test-user-2'))
        expect(isNicknameTaken).toBeTruthy()
      })

      it('should return false for available username', async () => {
        const isNicknameTaken = await store.dispatch(
          handlers.epics.isNicknameTaken('test-new-user')
        )
        expect(isNicknameTaken).toBeFalsy()
      })
    })

    describe('isNicknameTaken on empty store', () => {
      beforeEach(() => {
        store = create({
          initialState: {
            node: { ...NodeState, isTestnet: true },
            users: initialState
          }
        })
        jest.clearAllMocks()
      })

      it('should return false for available username when no users', async () => {
        const isNicknameTaken = await store.dispatch(handlers.epics.isNicknameTaken('test-user-2'))
        expect(isNicknameTaken).toBeFalsy()
      })
    })
  })
})
