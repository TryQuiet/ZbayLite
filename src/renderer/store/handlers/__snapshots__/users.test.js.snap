// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users reducer handles actions - setUsers sets users when empty 1`] = `
Object {
  "__altered": false,
  "__hash": undefined,
  "__ownerID": undefined,
  "_root": ArrayMapNode {
    "entries": Array [
      Array [
        "039ea1be861812ac6e210c96ac54fd2523f6d738a2db460613cc9aa8bd1886607f",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
      Array [
        "0214dc56ab3467e74b3bc1fed80d08e0146b251caceddc2ad6bfe37caca5fcc2e5",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname #2",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
      Array [
        "026918cbe1c9ace24ff065006c06b3c6344a5d7203ad42a45ddc7c7a41c6edfcc2",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname #3",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
    ],
    "ownerID": OwnerID {},
  },
  "size": 3,
}
`;

exports[`users reducer handles actions - setUsers sets users with same usernames 1`] = `
Object {
  "__altered": false,
  "__hash": undefined,
  "__ownerID": undefined,
  "_root": ArrayMapNode {
    "entries": Array [
      Array [
        "039ea1be861812ac6e210c96ac54fd2523f6d738a2db460613cc9aa8bd1886607f",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
      Array [
        "0214dc56ab3467e74b3bc1fed80d08e0146b251caceddc2ad6bfe37caca5fcc2e5",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname #2",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
      Array [
        "026918cbe1c9ace24ff065006c06b3c6344a5d7203ad42a45ddc7c7a41c6edfcc2",
        Object {
          "address": "ztestsapling14dxhlp8ps4qmrslt7pcayv8yuyx78xpkrtfhdhae52rmucgqws2zp0zwf2zu6qxjp96lzapsn4r",
          "createdAt": 1551963828,
          "firstName": "testname",
          "lastName": "testlastname",
          "nickname": "nickname #3",
          "onionAddress": "",
          "publicKey": undefined,
        },
      ],
    ],
    "ownerID": OwnerID {},
  },
  "size": 3,
}
`;
